src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -596,6 +596,17 @@ internal override CSharpAttributeData EarlyDecodeWellKnownAttribute(ref EarlyDec
                 return null;
             }
 
+            if (CSharpAttributeData.IsTargetEarlyAttribute(arguments.AttributeType, arguments.AttributeSyntax, AttributeDescription.AsyncMethodBuilderAttribute))
+            {
+                boundAttribute = arguments.Binder.GetAttribute(arguments.AttributeSyntax, arguments.AttributeType, out hasAnyDiagnostics);
+
+                var typeData = arguments.GetOrCreateData<CommonTypeEarlyWellKnownAttributeData>();
+                typeData.AsyncMethodBuilderTarget = boundAttribute.ConstructorArguments.FirstOrDefault();
+
+                // don't return the attribute. We need to bind it again later for full diagnostic reporting
```

Comment:
>// don't return the attribute. We need to bind it again later for full diagnostic reporting [](start = 16, length = 91)

It sounds like we don't check whether other attributes above (such as `AttributeUsageAttribute`, `ConditionalAttribute`, ...) are obsolete.
If that's the case, I wonder if this is on purpose and maybe we should not check whether `AsyncMethodBuilderAttribute` is obsolete either, so that we don't bind the attribute twice...
@alekseyts What do you think?

Reply:
@jcouv

> It sounds like we don't check whether other attributes above (such as AttributeUsageAttribute, ConditionalAttribute, ...) are obsolete.

Please verify and open an issue if that is the case. It appears that this function always returns null if there is any diagnostics, perhaps that takes care of the issue.

> If that's the case, I wonder if this is on purpose and maybe we should not check whether AsyncMethodBuilderAttribute is obsolete either, so that we don't bind the attribute twice...

It might be safe to return ```boundAttribute``` if ```!boundAttribute.HasErrors && !hasAnyDiagnostics```

---
In reply to: [232853589](https://github.com/dotnet/roslyn/pull/31021#discussion_r232853589) [](ancestors = 232853589)